### PR TITLE
v3: migration support for lookup vindexes

### PR DESCRIPTION
**Pull request by sougou**
Commits:
* change the vindex protocol allowing a non-unique vindex to return either a keyrange or keyspace ids.
* change lookup vindexes to support a new flag: "scatter_if_absent" that will cause it to return the full keyrange if an entry is not found. This behavior is changed in a newer commit.
* change the vindex protocol and requires Lookup vindexes to supply the Update function. This allow them to be more efficient, and some vindexes can choose to return an error if they cannot support it.
* add "autocommit" flag for lookup vindexes. Please see the commit comments for the associated behavior changes.
* change "scatter_if_absent" to "write_only".

### CODE REVIEW

**Comment by sougou** [comment]
@tirsen: This is phase 1, which is the hardest part of the feature.
The subsequent ones are:
* Create to be its own transaction.
* Option to upsert on Create.
* Disallow updates for such vindexes.

**Comment by slavikf** [comment]
75

**Comment by tirsen** [comment]
Remind me why upsert on create is needed?

**Comment by sougou** [comment]
upsert_on_create is for this use case:
* You're running in autocommit mode.
* You insert a row: Vindex insert succeeds but the actual insert fails.
* You attempt to insert the row again, but the Vindex will return a dup key error because the entry already exists.

With upsert_on_insert turned on, it will just replace the existing entry to match the new insert.

The downside:
You lose the ability to detect dup rows: If you had previously successfully completed an insert, then a new insert will replace the old vindex row making the previous row unreachable.

In other words, upsert_on_insert should be used only if you have other ways to guarantee uniqueness, and you want the ability to re-insert a previously failed insert.

File: go/vt/vtgate/engine/primitive.go [hunk 1]
```diff
@@ -42,10 +42,17 @@ const ListVarName = "__vals"
 type VCursor interface {
 	// Context returns the context of the current request.
 	Context() context.Context
+
+	// V3 functions.
```
Comment (demmer) [inline go/vt/vtgate/engine/primitive.go]: For the newcomer to the code it would be great to provide some more context on what "V3" means -- something like "Routing functions used by Vindex operations" would be clearer than "V3".

Perhaps you eventually plan to have the VCursor used in other contexts, but AFAIK `Execute` (and now `ExecuteAutocommit`) are only used by the vindex implementations so having this be clear in the interface would definitely help those of us who have tried to wade through what's going on here.

**Comment by demmer** [comment]
Some comments/questions more about the overall approach than the code specifics:

1. As discussed in Slack, the `scatter_if_absent` flag has a potential correctness issue for non-unique lookup vindexes during a backfill migration since it might get a partial but incomplete result set from the lookup table query. In other words, suppose a given non-unique vindex eventually will have two entries for a given column value, but in the middle of the migration a read operation is applied when only one of the rows has been written to the backing table. This wouldn't fall back to scatter since it's not an empty set, but it still isn't complete.

It seems to me the only way to handle a transitional non-unique lookup vindex is to always scatter on read until you are sure the vindex is fully populated. Note that the current behavior seems correct for unique lookup vindexes.

2. I'd love to understand the motivation for running vindex operations in their own transactions - although I _think_ I like this in general, but the complexities for autocommit are non-trivial as you point out here, and I haven't yet reasoned through how to properly sequence operations like secondary vindex updates or even keyspace id changes as mentioned in #3596.

3. Speaking of which, can you motivate why we want this flag to autocommit on insert? Is it performance driven? If so, one thing I've been thinking of adding is support to parse directives out of query comments, e.g. `INSERT /* multi-shard-autocommit=true  */ ...` would tell vitess to override the current logic and auto commit each row in a multi-row insert. This would of course violate ACID semantics but would have a performance benefit.

File: go/vt/vtgate/executor_dml_test.go [hunk 2]
```diff
@@ -531,6 +648,85 @@ func TestInsertSharded(t *testing.T) {
...
+			"type": "hash"
+		},
+		"name_user_map": {
+			"type": "lookup_hash",
+			"owner": "user",
+			"params": {
+				"table": "name_user_map",
+				"from": "name",
+				"to": "user_id",
+				"autocommit_on_insert": "true"
```
Comment (tirsen) [inline go/vt/vtgate/executor_dml_test.go]: It seems weird to use a string boolean value "true" when json supports boolean natively...

**Comment by tirsen** [comment]
@demmer 

"Autocommit" is for "transactional" integrity. That means we can't have a committed transaction that inserts a new row with a "lookup vindexed column" without the lookup table being updated since the lookup table insert has to commit before the transaction that inserts the vindexed column. It's a crucial part of "productionizing" vindexes. More details in my final comment here: https://github.com/youtube/vitess/issues/3537

Regarding atomicity of non-unique lookup tables:

I think if you need atomicity during the backfill you have to insert all the lookup rows for a given key in the same transaction. That can be accomplished using a map/reduce style backfill (where we shuffle to the target column value and then insert all of those in one transaction). We will backfill in our application code for now but I think this is how Vitess native backfills have to work to ensure correctness. I think Sugu has been toying with the idea of a Vitess native MR implementation, this may be a good first use case.

I think we can also very easily add a "silent vindex" option where we simply keep the lookup table up to date but don't use it for routing. This seems like a useful option if you know your backfill isn't atomic.

**Comment by sougou** [comment]
I signed it!

**Comment by sougou** [comment]
@demmer @rafael @tirsen this is now updated to match #3642.

**Comment by derekperkins** [comment]
@sougou Congrats on your first PR to the Vitess project!

File: go/vt/vtgate/vindexes/lookup_internal.go [hunk 3]
```diff
@@ -146,6 +177,9 @@ func (lkp *lookupInternal) Create(vcursor VCursor, rowsColValues [][]sqltypes.Va
 // A call to Delete would look like this:
 // Delete(vcursor, [[valuea, valueb]], 52CB7B1B31B2222E)
 func (lkp *lookupInternal) Delete(vcursor VCursor, rowsColValues [][]sqltypes.Value, value sqltypes.Value) error {
+	if lkp.Autocommit {
```
Comment (rafael) [inline go/vt/vtgate/vindexes/lookup_internal.go]: I think we should add a comment here, explaining why we are doing this.

**Comment by rafael** [comment]
I went couple times through this. LGTM.  Looking forward to having this throughly documented in vitess.io. It will be tricky to reason about all these different behaviors.

**Comment by alainjobart** [comment]
LGTM
